### PR TITLE
Add workspace flag to Railgun manager

### LINKED ISSUE #1229: KIC 2.0: configurable enterprise workspace

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:
- Adds a `kong-workspace` argument to the Railgun manager.
- Copies code from the 1.x cli utils into Railgun.
- Uses the current 1.x flag (`kong-admin-url`) to specify the admin URL, to avoid breaking existing configurations.

**Which issue this PR fixes**: fixes #1229 

**Special notes for your reviewer**:
- Importing the old main package into Railgun seemed wrong (and isn't possible), so I've copied code from there rather than splitting it into a new package. The code in question should go away soon anyway since it can be replaced with https://github.com/Kong/go-kong/pull/56 once that's released. Do we want to hold off on this until a go-kong release?
- I didn't see any reason that we were using the deprecated `--kong-url` flag in Railgun; that looked accidental. The workspace flag interacts with it, so I figured I'd align that with the current 1.x flag in this PR.

### CODE REVIEW

File: railgun/internal/mgrutils/workspace.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"context"
+	"fmt"
+	"github.com/kong/go-kong/kong"
+)
+
+// TODO these functions have been copied from 1.x cli/ingress-controller/util.go
+// https://github.com/Kong/go-kong/pull/56 provides go-kong replacements for them,
+// but is not yet released
```
Comment (mflendrich) [inline railgun/internal/mgrutils/workspace.go:11]: I have just cut go-kong v0.18.0 with that functionality included. Please go ahead and use that already.

File: railgun/manager/manager.go [hunk 2]
```diff
@@ -237,14 +240,26 @@ func Run(ctx context.Context, c *Config) error {
 		setupLog.Error(err, "cannot create a Kong Admin API client")
 	}
 
-	kongClient, err := kong.NewClient(&c.KongURL, httpclient)
+	kongClient, err := kong.NewClient(&c.KongAdminURL, httpclient)
 	if err != nil {
 		setupLog.Error(err, "unable to create kongClient")
 		return err
 	}
+	if c.KongWorkspace != "" {
```
Comment (mflendrich) [inline railgun/manager/manager.go:248]: To minimize the amount of logic in `main` and only perform wiring here, please wrap that in a function:

```go
func getKongClientForWorkspace(adminURL string, workspace string, httpclient *http.Client) (*kong.Client, error) {
  if workspace == "" {
    return kong.NewClient(&c.KongAdminURL, httpclient)
  }
  if err := /* check if workspace exists */; err != nil {
    return nil, fmt.Errorf("failed to ensure workspace in kong: %w", err)
  }
  return kong.NewClient(kong.String(c.KongAdminURL+"/"+c.KongWorkspace), httpclient)
}
```

and use it in a minimalistic fashion like this:

```go
kongClient, err := getKongClientForWorkspace(&c.KongAdminURL, &c.KongWorkspace, httpclient)
if err != nil {
  setupLog.Error(err, "unable to create kongClient")
  return err
}
```

File: railgun/manager/manager.go [hunk 3]
```diff
@@ -59,12 +59,13 @@ type Config struct {
...
 	LeaderElectionID     string
 	ProbeAddr            string
-	KongURL              string
+	KongAdminURL         string
 	FilterTag            string
 	Concurrency          int
 	KubeconfigPath       string
 	IngressClassName     string
 	AnonymousReports     bool
+	KongWorkspace        string
```
Comment (mflendrich) [inline railgun/manager/manager.go:68]: Please ensure that we test workspaces before we release beta.

That is: either add an acceptance criterion to #1279 that we test using workspaces, or create an issue on its own and add it to the milestone.

File: railgun/manager/manager.go [hunk 4]
```diff
@@ -97,12 +98,14 @@ func MakeFlagSetFor(c *Config) *pflag.FlagSet {
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
 	flagSet.StringVar(&c.LeaderElectionID, "election-id", "5b374a9e.konghq.com", `Election id to use for status update.`)
-	flagSet.StringVar(&c.KongURL, "kong-url", "http://localhost:8001", "TODO")
+	flagSet.StringVar(&c.KongAdminURL, "kong-admin-url", "http://localhost:8001",
```
Comment (mflendrich) [inline railgun/manager/manager.go:101]: Thanks for taking care of this! xref #1199

File: railgun/manager/manager.go (hunk 3 above)
Comment (rainest) [reply railgun/manager/manager.go:67]: Added along with the other Enterprise-specific flag (RBAC token).

File: railgun/manager/manager.go (hunk 2 above)
Comment (rainest) [reply railgun/manager/manager.go]: 2a5b053

File: pkg/adminapi/client.go [hunk 5]
```diff
@@ -71,3 +74,33 @@ func MakeHTTPClient(opts *HTTPClientOpts) (*http.Client, error) {
...
+
+// GetKongClientForWorkspace returns a Kong API client for a given root API URL and workspace.
+// If the workspace does not already exist, GetKongClientForWorkspace will create it.
+func GetKongClientForWorkspace(ctx context.Context, adminURL string, wsName string,
+	httpclient *http.Client) (*kong.Client, error) {
+	client, err := kong.NewClient(kong.String(adminURL), httpclient)
+	if err != nil {
+		return nil, fmt.Errorf("creating Kong client: %w", err)
+	}
+	if wsName != "" {
```
Comment (mflendrich) [inline pkg/adminapi/client.go:86]: If you invert this expression, there will be less indentation in this function, which improves code readability:

```go
if wsName == "" {
  // do the simple 1-liner thing
}

// do the complicated lengthy thing 1 level of nesting higher

```

File: railgun/manager/manager.go [hunk 6]
```diff
@@ -97,12 +97,14 @@ func MakeFlagSetFor(c *Config) *pflag.FlagSet {
...
+		`The Kong Admin URL to connect to in the format "protocol://address:port".`)
 	flagSet.StringVar(&c.FilterTag, "kong-filter-tag", "managed-by-railgun", "TODO")
 	flagSet.IntVar(&c.Concurrency, "kong-concurrency", 10, "TODO")
 	flagSet.StringVar(&c.KubeconfigPath, "kubeconfig", "", "Path to the kubeconfig file.")
 	flagSet.StringVar(&c.IngressClassName, "ingress-class", annotations.DefaultIngressClass, `Name of the ingress class to route through this controller.`)
 	flagSet.BoolVar(&c.AnonymousReports, "anonymous-reports", true, `Send anonymized usage data to help improve Kong`)
+	flagSet.StringVar(&c.KongWorkspace, "kong-workspace", "", `Kong Enterprise workspace to configure.`)
```
Comment (mflendrich) [inline railgun/manager/manager.go:107]:
```suggestion
	flagSet.StringVar(&c.KongWorkspace, "kong-workspace", "", `Kong Enterprise workspace to configure. " +
		"Leave this empty if not using Kong workspaces.")
```

**Review by mflendrich — APPROVED** [review]
LGTM assuming that all requested changes get applied

**Pull request merged**